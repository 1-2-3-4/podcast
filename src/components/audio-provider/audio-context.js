import * as React from 'react'

export const AudioContext = React.createContext({})