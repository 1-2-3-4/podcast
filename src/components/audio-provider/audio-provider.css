.loading {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: calc(100vh - 67px);
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
}

.loading-icon {
  -webkit-animation-name: spinner;
          animation-name: spinner;
  -webkit-animation-duration: 1s;
          animation-duration: 1s;
  -webkit-animation-iteration-count: infinite;
          animation-iteration-count: infinite;
  -webkit-animation-timing-function: linear;
          animation-timing-function: linear;
  width: 50px;
}

@-webkit-keyframes spinner {
  0% { -webkit-transform: rotate(0deg); transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); transform: rotate(360deg); }
}

@keyframes spinner {
  0% { -webkit-transform: rotate(0deg); transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); transform: rotate(360deg); }
}