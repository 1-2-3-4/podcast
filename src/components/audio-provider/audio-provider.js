import * as React from 'react'

import { AudioContext } from './audio-context'
import loadingIcon from './loading.svg'
import './audio-provider.css'

export const AudioProvider = ({ children }) => {
  const [trackData, setTrackData] = React.useState({})
  const [loading, setLoading] = React.useState(true)
  const [nowPlayingIndex, setNowPlayingIndex] = React.useState(0)

  React.useEffect(() => {
    const fetchAudioData = async () => {
      const { tracks } = await fetch('https://s3-us-west-2.amazonaws.com/anchor-website/challenges/bsb.json')
      .then(response => response.json()
        .then(data => data)
      )

      setTrackData(tracks)
      setLoading(false)
    }
    
    fetchAudioData()
  }, [setTrackData, setLoading]);

  if (loading) {
    return (
      <div className="loading">
        <img className="loading-icon" src={loadingIcon} alt="loading" />
      </div>
    )
  }

  return (
    <AudioContext.Provider
      value={{
        tracks: trackData,
        nowPlayingIndex,
        setNowPlayingIndex
      }}
    >
      {children}
    </AudioContext.Provider>
  )
}