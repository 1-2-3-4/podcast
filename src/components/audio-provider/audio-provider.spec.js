import * as React from 'react'
import { shallow } from 'enzyme'

import { AudioProvider } from './audio-provider'
import { AudioContext } from './audio-context'

describe('AudioProvider', () => {
  describe('when loading', () => {
    let wrapper

    beforeAll(() => {
      jest.spyOn(React, 'useEffect').mockImplementation(cb => cb())

      wrapper = shallow(<AudioProvider/>)
    })

    afterAll(() => {
      jest.restoreAllMocks()
    })

    it('renders loading spinner', () => {
      expect(wrapper.find({ className: 'loading' }).exists()).toBe(true)
    })

    it('matches snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })

  describe('when data has loaded', () => {
    let wrapper
    const mockTracks = [
      {
        title: 'This is a Fake Song',
        imageUrl: 'https://image.com/this-is-a-fake-song',
        mediaUrl: 'https://media.com/this-is-a-fake-song',
        duration: 10000
      },
      {
        title: 'Also a Mock Song',
        imageUrl: 'https://image.com/also-a-mock-song',
        mediaUrl: 'https://media.com/also-a-mock-song',
        duration: 10000
      },
      {
        title: 'Yup, Still More Fake Songs',
        imageUrl: 'https://image.com/yup-still-more-fake-songs',
        mediaUrl: 'https://media.com/yup-still-more-fake-songs',
        duration: 10000
      }
    ]

    beforeAll(() => {
      jest.spyOn(React, 'useEffect').mockImplementation(cb => cb())
      jest.spyOn(global, 'fetch').mockImplementation(jest.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve({ tracks: mockTracks }),
        })
      ))

      wrapper = shallow(<AudioProvider/>)
    })

    afterAll(() => {
      jest.restoreAllMocks()
    })

    it('renders provider with data', () => {
      expect(wrapper.find(AudioContext.Provider).prop('value')).toMatchObject({
        tracks: mockTracks,
        nowPlayingIndex: 0
      })
    })

    it('matches snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })
})