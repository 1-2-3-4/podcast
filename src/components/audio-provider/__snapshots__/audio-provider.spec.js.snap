// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioProvider when data has loaded matches snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "nowPlayingIndex": 0,
      "setNowPlayingIndex": [Function],
      "tracks": Array [
        Object {
          "duration": 10000,
          "imageUrl": "https://image.com/this-is-a-fake-song",
          "mediaUrl": "https://media.com/this-is-a-fake-song",
          "title": "This is a Fake Song",
        },
        Object {
          "duration": 10000,
          "imageUrl": "https://image.com/also-a-mock-song",
          "mediaUrl": "https://media.com/also-a-mock-song",
          "title": "Also a Mock Song",
        },
        Object {
          "duration": 10000,
          "imageUrl": "https://image.com/yup-still-more-fake-songs",
          "mediaUrl": "https://media.com/yup-still-more-fake-songs",
          "title": "Yup, Still More Fake Songs",
        },
      ],
    }
  }
/>
`;

exports[`AudioProvider when loading matches snapshot 1`] = `
<div
  className="loading"
>
  <img
    alt="loading"
    className="loading-icon"
    src="loading.svg"
  />
</div>
`;
